astropy/io/fits/tests/test_core.py

```diff
@@ -1388,3 +1388,20 @@ def test_error_if_memmap_impossible(self):
         # See https://github.com/astropy/astropy/issues/3766
         with fits.open(pth, memmap=True, do_not_scale_image_data=True) as hdul:
             hdul[0].data  # Just make sure it doesn't crash
+
+    def test_zero_byte_writes(self):
+
+        #use safeio which is a wrapper around python's io meant
+        #to intercept 0-byte writes
+        from astropy.io.tests import safeio
+
+        bright = np.rec.array([(1, 'Sirius', -1.45, 'A1V'),
+                            (2, 'Canopus', -0.73, 'F0Ib'),
+                            (3, 'Rigil Kent', -0.1, 'G2V')],
+                           formats='int16,a20,float32,a10', names='order,name,mag,Sp')
+
+        hdu_non_zero = fits.BinTableHDU(bright)
+        fh = safeio._BufferedWriter(open('bright.fits', mode='wb'))
+        hdu_non_zero.writeto(fh)
+        fh.close()
+        assert os.stat('bright.fits').st_size == fits.PrimaryHDU().filebytes() + hdu_non_zero.filebytes()
```

Comment:
```suggestion
        assert os.stat('bright.fits').st_size == fits.PrimaryHDU().filebytes() + hdu_non_zero.filebytes()

        hdu_img_2880 = fits.PrimaryHDU(data=np.arange(720, dtype='i4'))
        fh = safeio._BufferedWriter(open('image.fits', mode='wb'))
        hdu_img_2880.writeto(fh)
        fh.close()
        assert os.stat('image.fits').st_size == hdu_img_2880.filebytes()       
```
Padding in `_BaseHDU._writedata` is independent from the table padding and thus should get its own test.

Also I think the whole test belongs better into `TestFileFunctions` (one class above).